Refactors AsSeenFromMap to expose extension point.

paulp opened the pull request:
The extension point was the initial motivation but I also tried to
bring some clarity to the internals. This is a setup commit for
scaladoc and interactive modularization, and also will be followed
by a fix for abstract types losing their prefixes (SI-6161.)

paulp commented:
Review by @retronym, @adriaanm

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4345,147 +4345,146 @@ trait Types extends api.Types { self: SymbolTable =>
-          qvar.tpe
-      }
+    def apply(tp: Type): Type = tp match {
+      case ThisType(sym)            => thisTypeAsSeen(tp)
```
To save the match at the start of `thisTypeAsSeen`, sharpen its parameter type and:

``` scala
case tt @ ThisType(sym)            => thisTypeAsSeen(tt)
```

Ditto for `singleTypeAsSeen`.

retronym replied:
By the same token, you could make `classTypeParameterAsSeen` and `instParams` accept `TypeRef`s rather than `Type`s.

paulp replied:
It's not very satisfying in the end; I didn't do that because I still end up having redundant code somewhere. If I use the ClassTypeParameter extractor as is, I can't simultaneously promote it to a TypeRef; and in the case of instParams, I need to deconstruct the TypeRef anyway. I can access the members directly if the parameter is a typeref, but I don't consider it an improvement to be calling them tp2.sym, tp2.pre, tp2.args, instead of sym2, pre2, args2. If you can see a total-package improvement, can you describe it more completely?

retronym replied:
I would just prefer to use the more specific type as documentation/enforcement of what sort of type is expected at those places.

I think you can get the `TypeRef` out of `ClassTypeParameter` by changing the parameter type of `unapply`.

``` scala
scala> trait A; class B extends A; object E { def unapply(t: B) = Some((1, 2)) }
defined trait A
defined class B
defined module E

scala> def foo(a: A) = a match { case b @ E(_, _) => b; case _ => null }
foo: (a: A)B
```

Resuming a moderate level of typed extractor usage is an important milestone in the long rehabilitation process from SI-1697.

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4345,147 +4345,146 @@ trait Types extends api.Types { self: SymbolTable =>
+    private var _capturedSkolems: List[Symbol] = Nil
+    private var _capturedParams: List[Symbol]  = Nil
+    def capturedParams  = _capturedParams
+    def capturedSkolems = _capturedSkolems
```
I would give these two defs explicit return types, given that this class is to be an extension point.

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4345,147 +4345,146 @@ trait Types extends api.Types { self: SymbolTable =>
+    // 1) Walk the owner chain of `seenFromClass` until we find the class which owns `tp`
+    // 2) Take the base type of the prefix at that point with respect to the owning class
+    // 3) Solve for the type parameters through correspondence with the base type's type args
+    //
```
An example would help cement this comment.

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4345,147 +4345,146 @@ trait Types extends api.Types { self: SymbolTable =>
+    def capturedParams  = _capturedParams
+    def capturedSkolems = _capturedSkolems
+
+    def instParams(tp1: Type, tp2: Type): Type = {
```
Perhaps a better name? And you ought to move `// tp1: argument to AsSeenFromMap#apply` to a doc comment on this method.

retronym commented:
The meat of the patch looks great, a big improvement in readability.

As my last comment, I would be inclined to make all methods other than `instParams` final, unless they are intended as an extension points.

paulp committed:
Refactors AsSeenFromMap to expose extension point.

The extension point was the initial motivation but I also tried to
bring some clarity to the internals. This is a setup commit for
scaladoc and interactive modularization, and also will be followed
by a fix for abstract types losing their prefixes (SI-6161.)

paulp commented:
Another review by @retronym since this verges on a total rewrite.

paulp commented:
@adriaanm @retronym Where do we stand with this? If any of the identified reformulations are cause for alarm, I may need help constructing a case where it makes a difference.

adriaanm commented:
I thought traditionally PRs were guilty until proven innocent around here?

paulp commented:
Did I say otherwise?

adriaanm commented:
I was just trying to buy some time. Will try to have a look at the defense's remarks before the hour of noon.

paulp committed:
Simplified correspondingTypeArgument based on reviewer feedback.

adriaanm commented:
Ok, this is obviously a huge improvement in code clarity. Good stuff.

I'd like to ask for some of the review Q&A to be rendered as comments for posterity, and then it's good to go as far as I'm concerned. 

I worry a bit about the changes in widening prefixes. Is there a `tp` that's a valid prefix but for which `tp.widen.typeSymbol != tp.typeSymbol`?

retronym commented:
> Is there a tp that's a valid prefix but for which tp.widen.typeSymbol != tp.typeSymbol?

`scalacheck.Arbitrary[Type]` is sorely missed at times like these.

paulp commented:
I've already squeezed some fascinating info out of that question. Will report soon.

retronym commented:
Sorry that I've been in-absentia for the second half of this productive refactoring and review.

Would this be a good time to beef up tests for `asSeenFrom`? `run/compiler-asSeenFrom.scala` already pushes well into this area. But the intricate handling of `capturedSkolems` / `capturedExistentials` could benefit from some direct tests, too.

paulp commented:
Oh, I prayed nobody would ask for more tests. I started on that, but I'm not good at doing anything short of ludicrously exhaustive tests, and I kind of have a time problem.

Here's a prefix.typeSymbol vs. prefix.widen.typeSymbol teaser:

```
  abstract case class SuperType(thistpe: Type, supertpe: Type) extends SingletonType with SuperTypeApi {
    override def typeSymbol = thistpe.typeSymbol
    override def underlying = supertpe
  }
```

See that? "widen" is inherited from SimpleTypeProxy and forwards to "underlying", so supertpe. But typeSymbol is thistpe.

paulp committed:
Establishes what's up with widening in asSeenFrom.

Added explanatory comment.

paulp commented:
In the end, all the widening/not widening business in AsSeenFromMap is only there to, extremely indirectly, reconcile this. Surely nobody thought that by "widening" the type we were getting something which conforms to MORE types (did they?) but that is what is happening.

```
trait A { self: B => }
class B extends A with C
trait C

scala> rootMirror.staticClass("A") isSubClass rootMirror.staticClass("C")
res0: Boolean = false

scala> rootMirror.staticClass("A").thisType.widen.typeSymbol isSubClass rootMirror.staticClass("C")
res1: Boolean = true
```

Essentially .thisType.widen is our indirect route to .typeOfThis, but if anything stops at .thisType, everything in the self type is ignored. I can't even imagine how much this is responsible for.

```
scala> rootMirror.staticClass("A").tpe <:< rootMirror.staticClass("C").tpe
res9: Boolean = false

scala> rootMirror.staticClass("A").thisType <:< rootMirror.staticClass("C").tpe
res10: Boolean = true

scala> rootMirror.staticClass("A").thisType.widen <:< rootMirror.staticClass("C").tpe
res11: Boolean = true
```

paulp commented:
Wonderful... to perform the subclass check correctly, we have this beauty:

```
(pre.typeSymbol.typeOfThis.typeSymbol isSubClass clazz)
```

You might think that would be `pre.typeSymbol.thisSym` but you would think wrongly because of the "trick" performed by thisSym.

paulp commented:
Yep, apparently I'm not the first to discover this (unless this was also me, which sadly it might have been.)

```
val isProtectedOK = (
  (clasz.symbol isSubClass method.owner) ||
  (clasz.symbol.typeOfThis.typeSymbol isSubClass method.owner)  // see pos/bug780.scala
)
```

paulp commented:
Here's a "fun" alternate implementation of isSubClass.

```
    def isSubClass(that: Symbol): Boolean = isNonBottomSubClass(that) || {
      ((this ne thisSym) && (typeOfThis.typeSymbol isNonBottomSubClass that) && util.Origins("isSubClass") {
        println(s"Oops !($fullLocationString isSubClass ${that.fullLocationString}), but ${typeOfThis.typeSymbol.fullLocationString} is.")
        false
      })
    }
```

paulp commented:
Here are a few of the call sites which hit it.

```
>> Origins tag 'isSubClass' logged 7246 calls from 7 distinguished sources.

   2852   scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.hostForAccessorOf(SuperAccessors.scala:546)
   1770   scala.tools.nsc.typechecker.RefChecks$RefCheckTransformer.scala$tools$nsc$typechecker$RefChecks$RefCheckTransformer$$subOther$1(RefChecks.scala:347)
    949   scala.tools.nsc.backend.jvm.GenASM$JPlainBuilder$$anonfun$16.apply(GenASM.scala:1290)
    852   scala.tools.nsc.backend.jvm.GenASM$JPlainBuilder$$anonfun$minimizeInterfaces$1$1.apply(GenASM.scala:1292)
    734   scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:751)
     65   scala.tools.nsc.typechecker.RefChecks$RefCheckTransformer.scala$tools$nsc$typechecker$RefChecks$RefCheckTransformer$$subMember$1(RefChecks.scala:348)
     24   scala.tools.nsc.transform.ExplicitOuter.haveSameOuter(ExplicitOuter.scala:51)
```

adriaanm commented:
A's `self` should of course be considered to have type B, but I think it's expected that A is not considered a subclass/subtype of B. The knowledge that its self variable has type B does not leave the class body.

adriaanm commented:
So are you saying we need to widen A's ThisType to get to B?

paulp commented:
It is reasonable to expect that A is not a subclass of B, especially because B is a subclass of A. The problem is that "iSubClass" is used very frequently to mean "is this thing on the left also this thing on the right."

And that's right, we have to widen A's ThisType to get to B. How much code do you think has been written with the intuition that "A <:< B" can be true, but "A.typeSymbol isSubClass B.typeSymbol" false?

```
scala> rootMirror.staticClass("A").thisType <:< rootMirror.staticClass("B").tpe_*
res6: Boolean = true

scala> rootMirror.staticClass("A").thisType.typeSymbol isSubClass rootMirror.staticClass("B").tpe_*.typeSymbol
res7: Boolean = false
```

paulp commented:
@adriaanm I think this last commit brings in the necessary discussion, most of which is subsumed in the widening observation. I have to ask for a pass on any more tests right now.

adriaanm commented:
I'm okay  foregoing more tests. The most important thing is the widening I
think. Will have a look when I'm not restricted to thumb typing.

On Sunday, March 3, 2013, Paul Phillips wrote:

> @adriaanm https://github.com/adriaanm I think this last commit brings
> in the necessary discussion, most of which is subsumed in the widening
> observation. I have to ask for a pass on any more tests right now.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2163#issuecomment-14353147
> .

adriaanm commented:
LGTM
